cpp/ql/lib/semmle/code/cpp/models/implementations/Iterator.qll

```diff
@@ -108,91 +108,151 @@ private FunctionInput getIteratorArgumentInput(Operator op, int index) {
+  DataFlowFunction, TaintFunction {
   override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
-    input = iteratorInput and
+    input.isQualifierAddress() and
     output.isReturnValue()
     or
-    input.isParameterDeref(0) and output.isReturnValueDeref()
+    input.isReturnValueDeref() and
+    output.isQualifierObject()
+    or
+    input.isQualifierObject() and
+    output.isReturnValueDeref()
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isQualifierObject() and
+    output.isReturnValueDeref()
   }
 }
 
 /**
- * A non-member `operator+` function for an iterator type.
+ * A (member or non-member) `operator++` or `operator--` function for an iterator type.
  */
-private class IteratorAddOperator extends Operator, TaintFunction {
-  FunctionInput iteratorInput;
+class IteratorCrementOperator extends Function {
+  IteratorCrementOperator() {
+    this instanceof IteratorCrementNonMemberOperator or
+    this instanceof IteratorCrementMemberOperator
+  }
+}
 
-  IteratorAddOperator() {
+/**
+ * A non-member `operator+` function for an iterator type.
+ *
+ * Note that this class _only_ matches non-member functions. To find both
+ * non-member and member versions, use `IteratorBinaryAddOperator`.
+ */
+class IteratorAddNonMemberOperator extends Operator {
+  IteratorAddNonMemberOperator() {
     this.hasName("operator+") and
-    iteratorInput = getIteratorArgumentInput(this, [0, 1])
+    exists(getIteratorArgumentInput(this, [0, 1]))
   }
+}
 
+private class IteratorAddNonMemberOperatorModel extends IteratorAddNonMemberOperator, TaintFunction {
   override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
-    input = iteratorInput and
+    input = getIteratorArgumentInput(this, [0, 1]) and
     output.isReturnValue()
   }
 }
 
 /**
- * A non-member `operator-` function that takes a pointer difference type as its second argument.
+ * An `operator+` or `operator-` member function of an iterator class.
+ *
+ * Note that this class _only_ matches member functions. To find both
+ * non-member and member versions, use `IteratorBinaryAddOperator`.
  */
-private class IteratorSubOperator extends Operator, TaintFunction {
-  FunctionInput iteratorInput;
+class IteratorBinaryArithmeticMemberOperator extends MemberFunction {
+  IteratorBinaryArithmeticMemberOperator() {
+    this.getClassAndName(["operator+", "operator-"]) instanceof Iterator
+  }
+}
 
-  IteratorSubOperator() {
+private class IteratorBinaryArithmeticMemberOperatorModel extends IteratorBinaryArithmeticMemberOperator,
+  TaintFunction {
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isQualifierObject() and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * A (member or non-member) `operator+` or `operator-` function for an iterator type.
+ */
+class IteratorBinaryAddOperator extends Function {
```

Comment:
Based on the QLDoc, I think this should be called `IteratorBinaryArithmeticOperator` and should include `IteratorSubNonMemberOperator` (and `IteratorSubNonMemberOperator` should have the comment recommending use of this class instead).  Alternatively we could fully split the Add from the Sub operations.

Reply:
Well spotted. Fixed in fcc4c91739c9957106f19e9d85e3c42dd569d426.